src/dune_pkg/opam_solver.ml

```diff
@@ -0,0 +1,136 @@
+open Stdune
+
+module type CONTEXT = Opam_0install.S.CONTEXT
+
+(* Helper functor which implements [CONTEXT] for [(L.t, R.t) Either.t] where
+   [L] and [R] both implement [CONTEXT] *)
+module Context_either (L : CONTEXT) (R : CONTEXT) :
+  CONTEXT with type t = (L.t, R.t) Either.t = struct
+  type t = (L.t, R.t) Either.t
+
+  type rejection = (L.rejection, R.rejection) Either.t
+
+  let pp_rejection f = Either.map ~l:(L.pp_rejection f) ~r:(R.pp_rejection f)
+
+  let candidates =
+    let convert_rejections ~f =
+      List.map ~f:(fun (version, result) ->
+          (version, Result.map_error ~f result))
+    in
+    Either.map
+      ~l:(fun l name ->
+        L.candidates l name |> convert_rejections ~f:Either.left)
+      ~r:(fun r name ->
+        R.candidates r name |> convert_rejections ~f:Either.right)
+
+  let user_restrictions =
+    Either.map ~l:L.user_restrictions ~r:R.user_restrictions
+
+  let filter_deps = Either.map ~l:L.filter_deps ~r:R.filter_deps
+end
+
+(* Helper module for working with [OpamTypes.filtered_formula] *)
+module Filtered_formula : sig
+  open OpamTypes
+
+  (** Update a filtered formula setting the "with-test" and "with-doc" variables
+      to true. *)
+  val with_flags : Solver_env.Flag.Set.t -> filtered_formula -> filtered_formula
+end = struct
+  open OpamTypes
+
+  (* [map_formula_filters_bottom_up t ~f] calls [f] on each node of each filter
+     in t, bottom-up *)
+  let map_filters_bottom_up ~f =
+    OpamFilter.gen_filter_formula
+      (OpamFormula.partial_eval (function
+        | Filter flt -> `Formula (Atom (Filter (OpamFilter.map_up f flt)))
+        | Constraint _ as constraint_ -> `Formula (Atom constraint_)))
+
+  (* update a filtered formula replacing instances of a given variable with a given boolean value *)
+  let resolve_simple_bool_variable filtered_formula variable_to_resolve value =
+    map_filters_bottom_up filtered_formula ~f:(function
+      | FIdent ([], variable, None) as filter ->
+        if OpamVariable.equal variable variable_to_resolve then FBool value
+        else filter
+      | other -> other)
+
+  let with_flags flags filtered_formula =
+    Solver_env.Flag.Set.to_list flags
+    |> List.fold_left ~init:filtered_formula ~f:(fun acc flag ->
```

Comment:
You could just add a fold function to `Solver_env.Flag.Set`. It doesn't really matter much, but why favor list conversion when all you need is a fold?